src/stdlib_stringlist.f90

```diff
@@ -0,0 +1,726 @@
+! stdlib_stringlist.f90 --
+!     Module for storing and manipulating list of strings
+!     The strings may have arbitrary lengths, not necessarily the same
+!
+!     insert AT:      Inserts an element BEFORE the element present currently at the asked index
+!                       for forward indexes, otherwise
+!                     Inserts an element AFTER the element present currently at the asked index
+!                       for backward indexes
+!                     In other words, after insertion the element will be present at the asked index
+!                       for both forward and backward indexes                    
+!     insert BEFORE:  Inserts an element BEFORE the element present currently at the asked index
+!     insert AFTER:   Inserts an element AFTER the element present currently at the asked index
+!
+!     Note the distinction between AT and BEFORE in the module. Care has been taken to keep it consistent
+!     throughout the PR
+!
+module stdlib_stringlist
+    use stdlib_string_type, only: string_type, operator(/=) !, move
+    use stdlib_math, only: clip
+    ! use stdlib_optval, only: optval
+    implicit none
+    private
+
+    public :: stringlist_type, operator(//), operator(==), operator(/=)
+    public :: list_head, list_tail, fidx, bidx, stringlist_index_type
+
+    type stringlist_index_type
+        private
+        logical :: forward
+        integer :: offset
+
+    end type stringlist_index_type
+
+    type(stringlist_index_type), parameter :: list_head     = stringlist_index_type( .true. , 1 )   ! fidx(1)
+    type(stringlist_index_type), parameter :: list_tail     = stringlist_index_type( .false., 1 )   ! bidx(1)
+
+    !> Version: experimental
+    !> 
+    !> Returns an instance of type 'stringlist_index_type' representing forward index
+    !> [Specifications](../page/specs/stdlib_stringlist.html#fidx)
+    interface fidx
+        module procedure forward_index
+    end interface
+
+    !> Version: experimental
+    !> 
+    !> Returns an instance of type 'stringlist_index_type' representing backward index
+    !> [Specifications](../page/specs/stdlib_stringlist.html#bidx)
+    interface bidx
+        module procedure backward_index
+    end interface
+
+    type stringlist_type
+        private
+        integer :: size = 0
+        type(string_type), dimension(:), allocatable :: stringarray
+    
+    contains
+        private
+        procedure         :: copy                           =>  create_copy
+
+        procedure, public :: destroy                        =>  destroy_list
+
+        procedure, public :: len                            =>  length_list
+
+        procedure         :: capacity                       =>  capacity_list
```

Comment:
`len` and `capacity` seem like maybe they should be identical, but `len` doesn't check if the list is allocated.

Reply:
`capacity` is only for private use of developer.
whereas `len` is for the user using the module as well as the developer.
`capacity` is the length of the underlying array whereas `len` is the number of elements in the list. At this moment `len` and `capacity` behaves the same way but later if we decide to double the size of underlying array at insertion (instead of increasing the size by 1) then they will behave differently.

Reply:
In adherence to the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), I'd suggest only having one for now. At the very least, the public one should check the allocation status.

Reply:
If public one is checking the allocation status then should I remove the component `size` from the whole module?

Reply:
You probably could. I usually avoid keeping essentially duplicate information that needs to be kept consistent, as it's one more chance for bugs to creep in.

Reply:
What are thoughts on commit 1584eeb. Preserved `len` function but didn't remove `size` component from the module.
Removed some other lines as well.

Reply:
You should check the allocation status in the `len` function. I.e. the logic you had in the `capacity` function.